---
title: "2. Prediction_Model_Pipeline"
author: "Tyson Miller"
date: "2/7/2022"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(devtools)
library(broom)
library(data.table)
library(data.table)
```

Data from [here](https://uchicago.app.box.com/folder/102043737114) - genoGex.Rdata has everything we need in it
There are 5 'gex' RDS files which are the gene expressions for the 5 different tissues, the 'gtf' is the gene annotation, 'phyMap' is the snp annotation, and 'geno' is the genotype matrix

```{r}
load("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/genoGex.RData")
```

```{r}
#transposing gene expression files for the 5 tissues 
n = gexAc$EnsemblGeneID
gexAc_transpose <- as.data.frame(t(gexAc[,-1]))
colnames(gexAc_transpose) <- n

n = gexIl$EnsemblGeneID
gexIl_transpose <- as.data.frame(t(gexIl[,-1]))
colnames(gexIl_transpose) <- n

n = gexLh$EnsemblGeneID
gexLh_transpose <- as.data.frame(t(gexLh[,-1]))
colnames(gexLh_transpose) <- n

n = gexPl$EnsemblGeneID
gexPl_transpose <- as.data.frame(t(gexPl[,-1]))
colnames(gexPl_transpose) <- n

n = gexVo$EnsemblGeneID
gexVo_transpose <- as.data.frame(t(gexVo[,-1]))
colnames(gexVo_transpose) <- n
```

```{r}
# Running inverse normalization on each gene expression
invnorm = function(x) {
  if(is.null(dim(x))) res = invnorm.vector(x) else
  res=apply(x,2,invnorm.vector)
  res
}
invnorm.vector = function(x) {yy = rank(x)/(length(x)+1); qnorm(yy)}


gexAc_transpose = invnorm(gexAc_transpose)

gexIl_transpose = invnorm(gexIl_transpose)

gexLh_transpose = invnorm(gexLh_transpose)

gexPl_transpose = invnorm(gexPl_transpose)

gexVo_transpose = invnorm(gexVo_transpose)
```



```{r}
# Making the gene annotation file into the correct format for the pipeline
gtf$gene_type = sub(".*?gene_biotype(.*?);.*", "\\1", gtf$Attr)
gtf$gene_name = sub(".*?gene_name(.*?);.*", "\\1", gtf$Attr)

gene_annotation = subset(gtf, select = -c(Source, Feature, Score, Strand, Attr, Frame) )
gene_annotation = gene_annotation[, c("Chr","Gene", "gene_name", "Start", "End", "gene_type" )]
colnames(gene_annotation) = c("chr", "gene_id", "gene_name", "start", "end")
rownames(gene_annotation) = gene_annotation$gene_id
```

```{r}
# Making the snp annotation in the correct format for the pipeline
phyMap <- within(phyMap,  varID <- paste(Chr, Pos, Ref, Alt, sep="_"))
rownames(phyMap) = phyMap$varID
phyMap$rsid = phyMap$varID
colnames(phyMap) = c("snp", "chr", "pos", "refAllele", "effectAllele", 'varID', "rsid")
```

```{r}
# Splitting the snp annotation file by chromosome
s <- setNames(split(phyMap, phyMap$chr), paste0("snp_annot.chr", unique(phyMap$chr)))
list2env(s, globalenv())
```

```{r}
# writing the genotype file to a .txt file so that we can separate it by chromosome using our geneotype parse script.
rownames(geno) = rownames(phyMap)
write.table(geno, file = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/genotypes/genotype.txt", sep = "\t", col.names = TRUE, row.names = TRUE)
```

```{bash}
#Splitting the genotype file by chromosome - run this from the rat_genomic_alaysis directory
python scripts/split_genotype_by_chr.py genotypes/genotype.txt genotypes/geno_by_chr/'genotype'
```


```{r}
# Writing the gene expression files to csv files to be used for PEER Factor analysis
write.table(gexAc_transpose, file = '/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/gexAc.csv', sep = ",", col.names = TRUE, row.names = FALSE)

write.table(gexIl_transpose, file = '/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/gexIl.csv', sep = ",", col.names = TRUE, row.names = FALSE)

write.table(gexLh_transpose, file = '/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/gexLh.csv', sep = ",", col.names = TRUE, row.names = FALSE)

write.table(gexPl_transpose, file = '/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/gexPl.csv', sep = ",", col.names = TRUE, row.names = FALSE)

write.table(gexVo_transpose, file = '/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/gexVo.csv', sep = ",", col.names = TRUE, row.names = FALSE)
```

```{bash}
# Now we are performing PEER factor analysis on each tissue choosing 7 factors
peertool -f expression_files/"gexAc.csv" -n 7 -o peer_Ac --has_header

peertool -f expression_files/"gexIl.csv" -n 7 -o peer_Il --has_header

peertool -f expression_files/"gexLh.csv" -n 7 -o peer_Lh --has_header

peertool -f expression_files/"gexPl.csv" -n 7 -o peer_Pl --has_header

peertool -f expression_files/"gexVo.csv" -n 7 -o peer_Vo --has_header
```



```{r}
# Loading the phenotype file in to create covariate files. For this we are selecting sex, batch number, and batch center as covariates as well as the 7 PEER factors we generate
load("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/P50_raw_trait_values.RData")
covariatesAc = raw_traits[match(rownames(gexAc_transpose), raw_traits$rfid), ]
covariatesAc = subset(covariatesAc, select = c(rfid, sex, batchnumber, center))

covariatesIl = raw_traits[match(rownames(gexIl_transpose), raw_traits$rfid), ]
covariatesIl = subset(covariatesIl, select = c(rfid, sex, batchnumber, center))

covariatesLh = raw_traits[match(rownames(gexLh_transpose), raw_traits$rfid), ]
covariatesLh = subset(covariatesLh, select = c(rfid, sex, batchnumber, center))

covariatesPl = raw_traits[match(rownames(gexPl_transpose), raw_traits$rfid), ]
covariatesPl = subset(covariatesPl, select = c(rfid, sex, batchnumber, center))

covariatesVo = raw_traits[match(rownames(gexVo_transpose), raw_traits$rfid), ]
covariatesVo = subset(covariatesVo, select = c(rfid, sex, batchnumber, center))
```

```{r}
# Reading the PEER factor output files to be appended to the covariate file and eventually regressed out of the expression files
peer_factorsAc = read.csv(file = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/peer_Ac/X.csv", header = FALSE)

peer_factorsIl = read.csv(file = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/peer_Il/X.csv", header = FALSE)

peer_factorsLh = read.csv(file = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/peer_Lh/X.csv", header = FALSE)

peer_factorsPl = read.csv(file = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/peer_Pl/X.csv", header = FALSE)

peer_factorsVo = read.csv(file = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/peer_Vo/X.csv", header = FALSE)
```


```{r}
# Manipulating the PEER factor files so we can append to covariate file

colnames(peer_factorsAc) = c('PF1', 'PF2', 'PF3', 'PF4', 'PF5', 'PF6', 'PF7')
rownames(peer_factorsAc) = rownames(gexAc_transpose)

colnames(peer_factorsIl) = c('PF1', 'PF2', 'PF3', 'PF4', 'PF5', 'PF6', 'PF7')
rownames(peer_factorsIl) = rownames(gexIl_transpose)

colnames(peer_factorsLh) = c('PF1', 'PF2', 'PF3', 'PF4', 'PF5', 'PF6', 'PF7')
rownames(peer_factorsLh) = rownames(gexLh_transpose)

colnames(peer_factorsPl) = c('PF1', 'PF2', 'PF3', 'PF4', 'PF5', 'PF6', 'PF7')
rownames(peer_factorsPl) = rownames(gexPl_transpose)

colnames(peer_factorsVo) = c('PF1', 'PF2', 'PF3', 'PF4', 'PF5', 'PF6', 'PF7')
rownames(peer_factorsVo) = rownames(gexVo_transpose)
```


```{r}
#initializing matrices to be filled with t-stats, p_vals, and residuals of the regression of each gene vs. the covariates for each tissue. t-stat and p-val matrices are just for diagnostics

#t_statsAc = matrix(nrow = 13, ncol = length(colnames(gexAc_transpose)))
#p_valsAc = matrix(nrow = 13, ncol = length(colnames(gexAc_transpose)))
expressionAc = gexAc_transpose

#t_statsIl = matrix(nrow = 13, ncol = length(colnames(gexIl_transpose)))
#p_valsIl = matrix(nrow = 13, ncol = length(colnames(gexIl_transpose)))
expressionIl = gexIl_transpose

#t_statsLh = matrix(nrow = 13, ncol = length(colnames(gexLh_transpose)))
#p_valsLh = matrix(nrow = 13, ncol = length(colnames(gexLh_transpose)))
expressionLh = gexLh_transpose

#t_statsPl = matrix(nrow = 13, ncol = length(colnames(gexPl_transpose)))
#p_valsPl = matrix(nrow = 13, ncol = length(colnames(gexPl_transpose)))
expressionPl = gexPl_transpose

#t_statsVo = matrix(nrow = 13, ncol = length(colnames(gexVo_transpose)))
#p_valsVo = matrix(nrow = 13, ncol = length(colnames(gexVo_transpose)))
expressionVo = gexVo_transpose
```



```{r}
# Regressing out the covariates and saving the residuals as the new expression for each tissue
for (i in 1:length(colnames(gexAc_transpose))) {
    fit = lm(gexAc_transpose[,i] ~ covariatesAc$sex + covariatesAc$batchnumber + peer_factorsAc$PF1 + peer_factorsAc$PF2 + peer_factorsAc$PF3 + peer_factorsAc$PF4 + peer_factorsAc$PF5 + peer_factorsAc$PF6 + peer_factorsAc$PF7)
    expressionAc[,i] <- fit$residuals
    #t_statsAc[,i] <- tidy(fit)$statistic
    #p_valsAc[,i] <- tidy(fit)$p.value
}

for (i in 1:length(colnames(gexIl_transpose))) {
    fit = lm(gexIl_transpose[,i] ~ covariatesIl$sex + covariatesIl$batchnumber + peer_factorsIl$PF1 + peer_factorsIl$PF2 + peer_factorsIl$PF3 + peer_factorsIl$PF4 + peer_factorsIl$PF5 + peer_factorsIl$PF6 + peer_factorsIl$PF7)
    expressionIl[,i] <- fit$residuals
    #t_statsAc[,i] <- tidy(fit)$statistic
    #p_valsAc[,i] <- tidy(fit)$p.value
}

for (i in 1:length(colnames(gexLh_transpose))) {
    fit = lm(gexLh_transpose[,i] ~ covariatesLh$sex + covariatesLh$batchnumber + peer_factorsLh$PF1 + peer_factorsLh$PF2 + peer_factorsLh$PF3 + peer_factorsLh$PF4 + peer_factorsLh$PF5 + peer_factorsLh$PF6 + peer_factorsLh$PF7)
    expressionLh[,i] <- fit$residuals
    #t_statsAc[,i] <- tidy(fit)$statistic
    #p_valsAc[,i] <- tidy(fit)$p.value
}

for (i in 1:length(colnames(gexPl_transpose))) {
    fit = lm(gexPl_transpose[,i] ~ covariatesPl$sex + covariatesPl$batchnumber + peer_factorsPl$PF1 + peer_factorsPl$PF2 + peer_factorsPl$PF3 + peer_factorsPl$PF4 + peer_factorsPl$PF5 + peer_factorsPl$PF6 + peer_factorsPl$PF7)
    expressionPl[,i] <- fit$residuals
    #t_statsAc[,i] <- tidy(fit)$statistic
    #p_valsAc[,i] <- tidy(fit)$p.value
}

for (i in 1:length(colnames(gexVo_transpose))) {
    fit = lm(gexVo_transpose[,i] ~ covariatesVo$sex + covariatesVo$batchnumber + peer_factorsVo$PF1 + peer_factorsVo$PF2 + peer_factorsVo$PF3 + peer_factorsVo$PF4 + peer_factorsVo$PF5 + peer_factorsVo$PF6 + peer_factorsVo$PF7)
    expressionVo[,i] <- fit$residuals
    #t_statsAc[,i] <- tidy(fit)$statistic
    #p_valsAc[,i] <- tidy(fit)$p.value
  }
```

```{r}
# Save expression as tsv
Ac_expr <- as.data.frame(expressionAc) %>% mutate(ID = rownamse(expressionAc), .before = colnames(expressionAc))

Il_expr <- as.data.frame(expressionIl) %>% mutate(ID = rownamse(expressionIl), .before = colnames(expressionIl))

Lh_expr <- as.data.frame(expressionLh) %>% mutate(ID = rownamse(expressionLh), .before = colnames(expressionLh))

Pl_expr <- as.data.frame(expressionPl) %>% mutate(ID = rownamse(expressionPl), .before = colnames(expressionPl))

Vo_expr <- as.data.frame(expressionVo) %>% mutate(ID = rownamse(expressionVo), .before = colnames(expressionVo))


exprlist <- list(Ac_expr, Il_expr, Lh_expr, Pl_expr, Vo_expr)
tis <- c("Ac", "Il", "Lh", "Pl", "Vo")
i = 1
for(l in exprlist) {
write_tsv(l, "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/"
%&% tis[i] %&% "_expression_transformed.tsv", col_names = TRUE)
  i <- i+1
}
```


```{r}
# Saving the expression RDS objects to be used as arguments in the script
saveRDS(as.matrix(expressionAc), "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/Ac_expression_transformed.RDS")

saveRDS(as.matrix(expressionIl), "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/Il_expression_transformed.RDS")

saveRDS(as.matrix(expressionLh), "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/Lh_expression_transformed.RDS")

saveRDS(as.matrix(expressionPl), "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/Pl_expression_transformed.RDS")

saveRDS(as.matrix(expressionVo), "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/expression_files/Vo_expression_transformed.RDS")
```

```{r}
# Saving the SNP annotation RDS objects to be used as arguments in the script 
snp_files <- list(snp_annot.chr1, snp_annot.chr2, snp_annot.chr3, snp_annot.chr4, snp_annot.chr5, snp_annot.chr6, snp_annot.chr7, snp_annot.chr8, snp_annot.chr9, snp_annot.chr10, snp_annot.chr11, snp_annot.chr12, snp_annot.chr13, snp_annot.chr14, snp_annot.chr15, snp_annot.chr16, snp_annot.chr17, snp_annot.chr18, snp_annot.chr19, snp_annot.chr20)
i = 1
for(l in snp_files) {
  saveRDS(l, "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/genotypes/snp_annot/" %&% "snp_annot.chr" %&% i %&% ".RDS")
  i <- i+1
}
```

```{r}
# Saving the gene annotation RDS object to be used as an argument in the script
saveRDS(gene_annotation, "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/rat-genomic-analysis/genotypes/gene_annotation.RDS")
```



```{bash}
# Creating the meta data file for each tissue 
python scripts/create_meta_data.py --geno "genotypes/genotype.txt" --expr "expression_files/Ac_expression_transformed.tsv" --snpset "1KG" --rsid_label 1 --window 1000000 --out_prefix "Results/allMetaData/Ac"

python scripts/create_meta_data.py --geno "genotypes/genotype.txt" --expr "expression_files/Il_expression_transformed.tsv" --snpset "1KG" --rsid_label 1 --window 1000000 --out_prefix "Results/allMetaData/Il"

python scripts/create_meta_data.py --geno "genotypes/genotype.txt" --expr "expression_files/Lh_expression_transformed.tsv" --snpset "1KG" --rsid_label 1 --window 1000000 --out_prefix "Results/allMetaData/Lh"

python scripts/create_meta_data.py --geno "genotypes/genotype.txt" --expr "expression_files/Pl_expression_transformed.tsv" --snpset "1KG" --rsid_label 1 --window 1000000 --out_prefix "Results/allMetaData/Pl"

python scripts/create_meta_data.py --geno "genotypes/genotype.txt" --expr "expression_files/Vo_expression_transformed.tsv" --snpset "1KG" --rsid_label 1 --window 1000000 --out_prefix "Results/allMetaData/Vo"
```


```{bash}
# Running the model training script for each tissue/chromosome pair
for i in {1..20}
do
  Rscript scripts/create_model.R 'Ac' $i .5 1000000
  Rscript scripts/create_model.R 'Il' $i .5 1000000
  Rscript scripts/create_model.R 'Lh' $i .5 1000000
  Rscript scripts/create_model.R 'Pl' $i .5 1000000
  Rscript scripts/create_model.R 'Vo' $i .5 1000000
done
  
```


```{bash}
# Concatenating all of the results files for each tissue
bash scripts/make_all_results.sh 'Ac' 'Results/all_results_Ac' 0.5 '1KG_snps'
bash scripts/make_all_betas.sh 'Ac' 'Results/all_betas_Ac' 0.5 '1KG_snps'
bash scripts/make_all_logs.sh 'Ac' 'Results/all_logs_Ac'
bash scripts/make_all_covariances.sh 'Ac' 'Results/all_covariances_Ac' 0.5 '1KG_snps'

bash scripts/make_all_results.sh 'Il' 'Results/all_results_Il' 0.5 '1KG_snps'
bash scripts/make_all_betas.sh 'Il' 'Results/all_betas_Il' 0.5 '1KG_snps'
bash scripts/make_all_logs.sh 'Il' 'Results/all_logs_Il'
bash scripts/make_all_covariances.sh 'Il' 'Results/all_covariances_Il' 0.5 '1KG_snps'

bash scripts/make_all_results.sh 'Lh' 'Results/all_results_Lh' 0.5 '1KG_snps'
bash scripts/make_all_betas.sh 'Lh' 'Results/all_betas_Lh' 0.5 '1KG_snps'
bash scripts/make_all_logs.sh 'Lh' 'Results/all_logs_Lh'
bash scripts/make_all_covariances.sh 'Lh' 'Results/all_covariances_Lh' 0.5 '1KG_snps'

bash scripts/make_all_results.sh 'Pl' 'Results/all_results_Pl' 0.5 '1KG_snps'
bash scripts/make_all_betas.sh 'Pl' 'Results/all_betas_Pl' 0.5 '1KG_snps'
bash scripts/make_all_logs.sh 'Pl' 'Results/all_logs_Pl'
bash scripts/make_all_covariances.sh 'Pl' 'Results/all_covariances_Pl' 0.5 '1KG_snps'

bash scripts/make_all_results.sh 'Vo' 'Results/all_results_Vo' 0.5 '1KG_snps'
bash scripts/make_all_betas.sh 'Vo' 'Results/all_betas_Vo' 0.5 '1KG_snps'
bash scripts/make_all_logs.sh 'Vo' 'Results/all_logs_Vo'
bash scripts/make_all_covariances.sh 'Vo' 'Results/all_covariances_Vo' 0.5 '1KG_snps'
```

```{bash}
# Putting these into sql lite databases
python scripts/make_sqlite_db.py --output "Results/sql/Ac_output_db.db" --results "Results/all_results_Ac" --construction "Results/all_logs_Ac" --betas "Results/all_betas_Ac" --meta "Results/allMetaData/Ac.allMetaData.txt"

python scripts/make_sqlite_db.py --output "Results/sql/Il_output_db.db" --results "Results/all_results_Il" --construction "Results/all_logs_Il" --betas "Results/all_betas_Il" --meta "Results/allMetaData/Il.allMetaData.txt"

python scripts/make_sqlite_db.py --output "Results/sql/Lh_output_db.db" --results "Results/all_results_Lh" --construction "Results/all_logs_Lh" --betas "Results/all_betas_Lh" --meta "Results/allMetaData/Lh.allMetaData.txt"

python scripts/make_sqlite_db.py --output "Results/sql/Pl_output_db.db" --results "Results/all_results_Pl" --construction "Results/all_logs_Pl" --betas "Results/all_betas_Pl" --meta "Results/allMetaData/Pl.allMetaData.txt"

python scripts/make_sqlite_db.py --output "Results/sql/Vo_output_db.db" --results "Results/all_results_Vo" --construction "Results/all_logs_Vo" --betas "Results/all_betas_Vo" --meta "Results/allMetaData/Vo.allMetaData.txt"
```



