---
title: "3.PrediXcan_MultiXcan"
author: "Natasha Santhanam"
date: "2/7/2022"
output: html_document
---

```{r setup, eval=FALSE}
library(tidyverse)
library(devtools)
library(broom)
library(data.table)
library(RSQLite)
library(data.table)
library(qqman)
devtools::source_gist("0ddc9c0ea03245bb30efbe3e899897be")
"%&%" = function(a,b) paste(a,b,sep="")
args = commandArgs(trailingOnly=TRUE)
geno.dir <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/rat_genotypes_LD_pruned_0.95/"
```

## Run PrediXcan with Metabolic Phenotype Data

# Create model with right column names for PrediXcan (do for all 5 tissues)
```{r create txt genotypes, eval=FALSE}
filelist <- list.files(geno.dir, pattern = ".bimbam")
#ids for rats are in the phenotype file under rat_rfid

for(fila in filelist) {
  tempo <- fread(geno.dir %&% fila)
  tempo <- tempo %>% mutate(chr =  numextract(sapply(strsplit(tempo$V1, ":"), `[`, 1)), .before = V1) %>% mutate(pos = numextract(sapply(strsplit(tempo$V1, ":"), `[`, 2)), .before = V2) %>% mutate(maf = 0, .before = V4)
  write_tsv(tempo, geno.dir %&% substr(fila, 1, nchar(fila) - 7) %&% ".txt", col_names = FALSE)
}
```


```{r change colnames of Ac prediction model, eval=FALSE}
filename <- MODEL %&% "Ac_output_db.db"
  sqlite.driver <- dbDriver("SQLite")
  conn <- dbConnect(RSQLite::SQLite(), filename)
  extra <- dbGetQuery(conn, 'select * from extra')
  weights <- dbGetQuery(conn, 'select * from weights')
extra <- extra %>% select(c(gene, genename, n.snps, R2, pval)) %>% mutate(pred.perf.qval = NA) 
colnames(extra) <- c("gene", "genename", "n.snps.in.model", "pred.perf.R2", "pred.perf.pval", "pred.perf.qval")
```


```{r create database connection, eval=FALSE}
model_db = MODEL %&% "Ac_annot_prediXcan_db.db"
conn <- dbConnect(RSQLite::SQLite(), model_db)
dbWriteTable(conn, "weights", weights)
dbWriteTable(conn, "extra", extra)

#check to see model is set up 
dbListTables(conn)
dbGetQuery(conn, 'SELECT * FROM weights') %>% head
dbGetQuery(conn, 'SELECT * FROM extra') %>% head
dbDisconnect(conn)
```

# Do for all 5 tissues 
```{bash run prediXcan, eval = FALSE}
#run prediXcan
conda activate /gpfs/data/im-lab/nas40t2/bin/envs/tensorqtl/ 
export METAXCAN=/gpfs/data/im-lab/nas40t2/natasha/GTEX_Analysis/MetaXcan/software
export GENO=/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/rat_genotypes_LD_pruned_0.95
export MODEL=/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/models

python $METAXCAN/Predict.py \
--model_db_path $MODEL/Ac_annot_prediXcan_db.db \
--text_genotypes  \
 $GENO/chr*.round2_impute2_3473.txt  \
--on_the_fly_mapping METADATA "{}_{}_{}_{}" \
--text_sample_ids $GENO/samples_Rat_metab_phenos_file \
--prediction_output rat_metabolic_Ac__predict.txt  \
--prediction_summary_output rat_metabolic_Ac__summary.txt \
--throw
```

```{r run associations, eval=FALSE}
#run asssociation in prediXcan
PHENO = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/"
RESULTS = "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/prediXcan/metabolic_traits/"

pheno <- read_csv(PHENO %&% "processed_obesity_rat_Palmer_phenotypes.csv", col_names = TRUE)

for(i in 2:length(colnames(pheno))){
trait <- colnames(pheno)[i]
runLOC <- "python3 " %&% METAXCAN %&% "/PrediXcanAssociation.py " %&% "--expression_file " %&% RESULTS %&% "rat_metabolic_" %&% args[1] %&% "__predict.txt --input_phenos_file " %&% PHENO %&% "processed_obesity_rat_Palmer_phenotypes.tsv " %&% "--input_phenos_column " %&% trait %&%  " --output " %&% RESULTS %&% "associations/" %&% args[1] %&% "__association_" %&% trait %&% ".txt --verbosity 9 --throw"
system(runLOC)
}
```


```{bash submit script for all tissues}
Rscript --vanilla /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/prediXcan/metabolic_trait_assoc_all_tis.R $tissue

qsub -v tissue=$tis metabolic_assoc_all_tissues.pbs
```

Format and Plot PrediXcan Results

```{r format prediXcan assocs, eval=FALSE}
results.dir <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/prediXcan/associations/"
filelist <- list.files(results.dir, pattern = "mice_Palmer__association", full.names = TRUE)
full_df <- data.frame()

for(fila in filelist) {
  assoc_fila <- read_tsv(fila, col_names = TRUE)
  pheno_id <- substr(fila, 97, (str_length(fila) - 5))
  tempo <- cbind(assoc_fila, metabolic_trait=pheno_id) %>% select(-c(status))
  full_df <- rbind(full_df, tempo)
} 
#full_df <- read_tsv("/Users/natashasanthanam/Github/rat-genomic-analysis/data/rat_metabolic_traits_Ac_full_assocs.txt", col_names = TRUE)

tempo_df <- full_df %>% filter(pvalue < 5.836349e-06)

#589 sig genes
tempo_df %>% group_by(gene) %>% summarise(n = n())

#all 10 traits
tempo_df %>% group_by(metabolic_trait) %>% summarise(n = n())
```


```{r find genes that overlap in Humans, eval=FALSE}
human_height_genes <- readRDS("/Users/natashasanthanam/Downloads/MultiXcan_zscores_all_species.RDS") %>% mutate(rat_gene = orth.rats[match(human_height_genes$gene_name, orth.rats$external_gene_name), 4]$rnorvegicus_homolog_associated_gene_name) %>% filter(abs(Height) >= 3.719 & abs(zscore.bodylength_w_tail) >= 2.32)


human_bmi_genes <- readRDS("/Users/natashasanthanam/Downloads/MultiXcan_zscores_all_species.RDS") %>% mutate(rat_gene = orth.rats[match(human_bmi_genes$gene_name, orth.rats$external_gene_name), 4]$rnorvegicus_homolog_associated_gene_name) %>% filter(abs(BMI) >= 3.719 & abs(zscore.bmi_bodylength_w_tail) >= 2.32)
```


```{r plot prediXcan results as miami plot divided per trait}
gene_annot <- readRDS("/Users/natashasanthanam/Github/rat-genomic-analysis/data/gene_annotation.RDS") %>% select(c("chr", "gene_id", "start", "end")) %>% rename(gene = gene_id)

tempo_manhatt <- inner_join(gene_annot, full_df, by = "gene")
tempo_manhatt$chr <- as.numeric(tempo_manhatt$chr)

bmi_manhat <- tempo_manhatt %>% filter(metabolic_trait == "bmi_bodylength_w_tail") %>% mutate(gene_name = orth.rats[match(bmi_manhat$gene, orth.rats$rnorvegicus_homolog_ensembl_gene), 4]$rnorvegicus_homolog_associated_gene_name)

height_manhat <- tempo_manhatt %>% filter(metabolic_trait == "bodylength_w_tail") %>% mutate(gene_name = orth.rats[match(height_manhat$gene, orth.rats$rnorvegicus_homolog_ensembl_gene), 4]$rnorvegicus_homolog_associated_gene_name)
```

```{r plot manhattan for BMI}
data_cum <- bmi_manhat %>% 
  group_by(chr) %>% 
  summarise(max_bp = as.numeric(max(start))) %>% 
  mutate(bp_add = lag(cumsum(max_bp), default = 0)) %>% 
  select(chr, bp_add)

gwas_data <- bmi_manhat %>% 
  inner_join(data_cum, by = "chr") %>% 
  mutate(bp_cum = start + bp_add)


axis_set <- gwas_data %>% 
  group_by(chr) %>% 
  summarize(center = mean(bp_cum))

ylim <- gwas_data %>% 
  filter(pvalue == min(pvalue)) %>% 
  mutate(ylim = abs(floor(log10(pvalue))) + 2) %>% 
  pull(ylim)

sig <- 5.767678e-07

manhplot <- ggplot(gwas_data, aes(x = bp_cum, y = -log10(pvalue), 
                                  color = as_factor(chr), size = -log10(pvalue))) +
  geom_hline(yintercept = -log10(sig), color = "grey40", linetype = "dashed") + 
  geom_hline(yintercept = -log10(0.0001), color = "red", linetype = "dashed") + 
  geom_point(alpha = 0.75, shape = ifelse((gwas_data$zscore >= 4.863456), 17, ifelse(gwas_data$zscore <= -4.863456, 25, 19)), fill = "dodgerblue") +
  geom_text_repel(aes(label=ifelse(pvalue <=  sig, gene_name, ""))) + 
  scale_x_continuous(label = axis_set$chr, breaks = axis_set$center) +
  scale_y_continuous(expand = c(0,0), limits = c(0, ylim)) +
  scale_color_manual(values = rep(c("#276FBF", "#183059"), unique(length(axis_set$chr)))) +
  scale_size_continuous(range = c(0.5,3)) +
  labs(x = NULL, 
       y = expression(-log[10](italic(p)))) + 
  theme_minimal() +
  theme( 
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(angle = 60, size = 8, vjust = 0.5)
  )
```


```{r plot manhattan for height}
data_cum <- height_manhat %>% 
  group_by(chr) %>% 
  summarise(max_bp = as.numeric(max(start))) %>% 
  mutate(bp_add = lag(cumsum(max_bp), default = 0)) %>% 
  select(chr, bp_add)

gwas_data <- height_manhat %>% 
  inner_join(data_cum, by = "chr") %>% 
  mutate(bp_cum = start + bp_add)


axis_set <- gwas_data %>% 
  group_by(chr) %>% 
  summarize(center = mean(bp_cum))

ylim <- gwas_data %>% 
  filter(pvalue == min(pvalue)) %>% 
  mutate(ylim = abs(floor(log10(pvalue))) + 2) %>% 
  pull(ylim)

sig <- 5.767678e-07

manhplot <- ggplot(gwas_data, aes(x = bp_cum, y = -log10(pvalue), 
                                  color = as_factor(chr), size = -log10(pvalue))) +
  geom_hline(yintercept = -log10(sig), color = "grey40", linetype = "dashed") + 
  geom_hline(yintercept = -log10(0.0001), color = "red", linetype = "dashed") + 
  geom_point(alpha = 0.75, shape = ifelse((gwas_data$zscore >= 4.863456), 17, ifelse(gwas_data$zscore <= -4.863456, 25, 19)), fill = "dodgerblue") +
  geom_text_repel(aes(label=ifelse(pvalue <=  sig, gene_name, ""))) + 
  scale_x_continuous(label = axis_set$chr, breaks = axis_set$center) +
  scale_y_continuous(expand = c(0,0), limits = c(0, ylim)) +
  scale_color_manual(values = rep(c("#276FBF", "#183059"), unique(length(axis_set$chr)))) +
  scale_size_continuous(range = c(0.5,3)) +
  labs(x = NULL, 
       y = expression(-log[10](italic(p)))) + 
  theme_minimal() +
  theme( 
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(angle = 60, size = 8, vjust = 0.5)
  )

```



## Run MultiXcan

# Generate Folder with Predicted Expression Data for each Tissue

First have to remove potential overlap between genotypes used in predicted expression and those in phenotype file. Should only be around 60ish so not too big a deal

```{r generate list of ids across all tissues, eval=FALSE}
filelist <- list.files("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/expression", pattern = ".RDS", full.names = TRUE)
all_names <- data.frame(ID = as.character())
for(fila in filelist) {
  tempo <- readRDS(fila)
  tempo <- as.data.frame(rownames(tempo)) %>% rename(ID = `rownames(tempo)`)
  all_names <- full_join(tempo, all_names, by = "ID")
}
```


```{r clean up pheno file, eval=FALSE}
pheno <- read_csv("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/processed_obesity_rat_Palmer_phenotypes.csv", col_names=TRUE)

pheno <- pheno %>% rename(ID = rat_rfid) %>% filter(!ID %in% all_names$ID)
```


Next have to remove overlap rats predicted expression as well
```{r  generate expr data, eval=FALSE}
filelist <- list.files("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/prediXcan", pattern = "__predict.txt", full.names = TRUE)

for(fila in filelist) {
  tempo <- fread(fila, header=TRUE)
  name <- substr(fila, 72,73)
  tempo <- tempo %>% filter(!FID %in% all_names$ID)
  tempo <- tempo[match(pheno$ID, tempo$FID),]
  write_tsv(tempo, "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/MultiXcan/expr/" %&% name %&% ".txt")
}
```


Run MultiXcan using the predicted expression from prediXcan across all 5 tissues to boost power
```{bash run MultiXcan}
#!/bin/bash

#PBS -N multixcan
#PBS -S /bin/bash
#PBS -l walltime=4:00:00
#PBS -l mem=4gb
#PBS -l nodes=1:ppn=1

# SPECIFY LOGGING BEHAVIOR

#PBS -o /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/MultiXcan/logs/${PBS_JOBNAME}.${PBS_JOBID}.log
#PBS -e /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/MultiXcan/logs/${PBS_JOBNAME}.${PBS_JOBID}.err

module load gcc/6.2.0
source ~/.bashrc 
conda activate /gpfs/data/im-lab/nas40t2/bin/envs/tensorqtl/

echo "MultiXcan running on epifat"

python /gpfs/data/im-lab/nas40t2/natasha/GTEX_Analysis/MetaXcan/software/MulTiXcan.py \
        --expression_folder /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/MultiXcan/expr \
        --expression_pattern "(.*)_expression_transformed.txt" \
        --input_phenos_file /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/MultiXcan/metabolic_trait_phenos_MultiXcan.txt \
        --input_phenos_column fasting_glucose \
        --output /gpfs/data/im-lab/nas40t2/natasha/rat_genomics/MultiXcan/results/fasting_glucose_predict_assoc.txt \
        --pc_condition_number 10 \
        --mode linear \
        --verbosity 8 \
        --throw
```


# Add Zscore to MutliXcan

Calculate most significant  Zscore across all tisuses
For each trait find most significant pvalue and take sign of that effect

```{r find mean Z, eval=FALSE}
pheno <- read_csv("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/Box_files/processed_obesity_rat_Palmer_phenotypes.csv", col_names=TRUE)

for(i in 9:ncol(pheno)) {
  trait <- colnames(pheno)[i]
  filelist <- list.files(data.dir %&% "prediXcan/metabolic_traits/associations/", pattern = trait %&% ".txt", full.names = TRUE)
  tempo <- data.frame(gene= as.character())
  for(fila in filelist) {
    tis <- substr(fila, 89, 90)
    df <- read_tsv(fila) %>% select(c(gene, effect, pvalue)) 
    new_eff <- paste("effect", tis, sep = "_")
    new_pval <- paste("pvalue", tis, sep = "_")
    colnames(df)[2] <- new_eff
    colnames(df)[3] <- new_pval 
    
   tempo <- full_join(tempo, df,  by = "gene")
  }
  most_sig = rowMins(as.matrix(tempo[,c(3,5,7,9,11)]))
  
  Ac <- tempo[na.omit(match(most_sig, tempo$pvalue_Ac)), c(1,2)] %>% rename(effect = effect_Ac )
  Il <- tempo[na.omit(match(most_sig, tempo$pvalue_Il)), c(1,4)] %>% rename(effect = effect_Il )
  Lh <- tempo[na.omit(match(most_sig, tempo$pvalue_Lh)), c(1,6)] %>% rename(effect = effect_Lh )
  Pl <- tempo[na.omit(match(most_sig, tempo$pvalue_Pl)), c(1,8)] %>% rename(effect = effect_Pl )
  Vo <- tempo[na.omit(match(most_sig, tempo$pvalue_Vo)), c(1,10)] %>% rename(effect = effect_Vo )
  
  df <- rbind(Ac, Il, Lh, Pl, Vo) 
  df <- df %>% mutate(sign = sign(effect))
  write_tsv(df, data.dir %&% "prediXcan/metabolic_traits/associations/most_sig_zscores/" %&% trait %&% "_avg_zscore.txt", col_names = FALSE)
}
```

