---
title: "04.PhenomeXcan_MultiXcan.Rmd"
author: "Natasha Santhanam"
date: "2/14/2022"
output: html_document
---

---
title: "PhenomeXcan Query"
author: "Natasha Santhanam"
date: "10/12/2021"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(devtools)
library(broom)
library(data.table)
library(RSQLite)
library(qqman)
library(ggrepel)
library(ggpubr)
"%&%" = function(a,b) paste(a,b,sep="")
data.dir <- "/Users/natashasanthanam/Github/rat-genomic-analysis/"
```

```{r}
suppressPackageStartupMessages(source(data.dir %&% "helpers.R", chdir = TRUE))
phenomexcan_con <- get_db()
dbListTables(phenomexcan_con)
```

## query PhenomeXcan association with top phenotypes for list of genes


```{r query PhenomeXcan for top genes associated with Obesity }
input = list()
input$pheno = c("Obesity")
input$limit = 30000
obesity_genes <-  suppressMessages(get_results_from_data_db(input))
```


```{r query PhenomeXcan for top genes associated with Body fat percentage }
input = list()
input$pheno = c("Body fat percentage")
input$limit = 30000
body_fat_genes <- suppressMessages(get_results_from_data_db(input))
```

```{r query PhenomeXcan for top genes associated with BMI }
input = list()
input$pheno = c("Body mass index (BMI) (21001_raw)")
input$limit = 30000
BMI_genes <- suppressMessages(get_results_from_data_db(input))
```


```{r query PhenomeXcan for top genes associated with Fasting Glucose }
input = list()
input$pheno = c("Fasting Glucose")
input$limit = 30000
glucose_genes <- suppressMessages(get_results_from_data_db(input))
```


```{r query PhenomeXcan for top genes associated with Height }
input = list()
input$pheno = c("Height")
input$limit = 30000
height_genes <- suppressMessages(get_results_from_data_db(input))
```


Generate a table of all human MultiXcan results
```{r create phenotype matrix in Humans}
#matrix - humans (rows are genes and columns are traits (fat, BMI, Obesity))
listphenos <- list(BMI_genes, body_fat_genes, obesity_genes, glucose_genes, height_genes)
pheno_Multi_humans <- data_frame(gene_name = as.character())

for(l in listphenos) {
  trait <- l$phenotype[1]
  tempo <- l %>% dplyr::select(c(gene_name, pvalue))
  colnames(tempo)[2] = paste("pvalue", trait, sep="_")
  pheno_Multi_humans <- full_join(pheno_Multi_humans, tempo, by = "gene_name")
}


human_genes <- as.data.frame(pheno_Multi_humans$gene_name)

#pheno_humans <- as.matrix(pheno_humans %>% dplyr::select(-c(gene_name)))
```


```{r find orthologs in rats}
human = useEnsembl(biomart='ensembl', dataset="hsapiens_gene_ensembl", mirror = "uswest")
#human = useMart("ensembl", dataset = "hsapiens_gene_ensembl", mirror = "uswest")
attributes <- listAttributes(human)

attributes = c("ensembl_gene_id", "external_gene_name", "rnorvegicus_homolog_ensembl_gene", "rnorvegicus_homolog_associated_gene_name")
orth.rats = getBM(attributes, filters="with_rnorvegicus_homolog",values=TRUE, mart = human, uniqueRows=TRUE)

human_genes <- human_genes %>% rename(external_gene_name = `pheno_Multi_humans$gene_name`)
human_genes <- inner_join(human_genes, orth.rats, by = "external_gene_name") %>% dplyr::select(c(external_gene_name, rnorvegicus_homolog_associated_gene_name))
```

Generate a table of all rat MultiXcan results
```{r make rat multiXcan matrix, eval=FALSE}
filelist <- list.files("/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/MultiXcan/results", pattern = "assoc.txt",full.names = TRUE)

pheno_Multi_rat<- data_frame(gene = as.character())
for(fila in filelist) {
  trait <- substr(fila, 66, str_length(fila)-18)
  tempo <- fread(fila) %>% select(c(gene, pvalue))
  colnames(tempo)[2] = paste("pvalue", trait, sep=".")
  pheno_Multi_rat<- full_join(pheno_Multi_rat, tempo, by = "gene")
}

pheno_Multi_rat <- read_tsv("/Users/natashasanthanam/Downloads/rat_metabolic_MultiXcan_pval_lasso_assoc.txt", col_names = TRUE)

pheno_Multi_rat <- pheno_Multi_rat %>% mutate(gene_name = orth.rats[match(pheno_Multi_rat$gene, orth.rats$rnorvegicus_homolog_ensembl_gene),2]$external_gene_name)
```


# Enrichment Analysis
```{r enrichment for fasting glucose}
devtools::source_gist("38431b74c6c0bf90c12f")
devtools::source_gist("1e9053c8f35c30396429350a08f33ea7")
```

```{r enrichment for BMI and Height}
full_df <- full_df %>% mutate(gene_name = orth.rats[match(full_df$gene, orth.rats$rnorvegicus_homolog_ensembl_gene), 2]$external_gene_name)

qqunif(BMI_genes$pvalue, col= "maroon1" )
BMI_genes_rats <- full_df %>% filter(metabolic_trait == "bmi_bodylength_wo_tail")  %>% filter(pvalue <= 0.05) %>% .[["gene_name"]]
qqpoints(BMI_genes %>% filter(gene_name %in% BMI_genes_rats) %>% .[["pvalue"]],pch='+', col = "dodgerblue") 


qqunif(height_genes$pvalue, col = "maroon1")
height_genes_rats <- full_df %>% filter(metabolic_trait == "bodylength_w_tail") %>% filter(pvalue <= 0.05) %>% .[["gene_name"]]
qqpoints(height_genes %>% filter(gene_name %in% height_genes_rats) %>% .[["pvalue"]],pch='+', col="dodgerblue")
```


#Make enrichment plot but flip rats and humans

Make three points
```{r flipped enrichment plot in rasts for fasting glucose}
qqunif(pheno_rats$fasting_glucose,main="Fasting Glucose Enrichment Plot in Rats")
fasting_glucose_humans <- glucose_genes %>% filter(pvalue <= 0.01 && rcp >= 0.01) %>% .[["gene_name"]] %>% na.omit() 
qqpoints(pheno_rats %>% filter(gene_name %in% fasting_glucose_humans ) %>% .[["fasting_glucose"]],pch='+', col=3)
```


```{r flipped enrichment plot for epifat}
qqunif(pheno_rats$epifat,main="Epfiat  Enrichment Plot in Rats")
body_fat_humans <- body_fat_genes %>% filter(pvalue <= 0.01 && rcp >= 0.01) %>% .[["gene_name"]] %>% na.omit() 
qqpoints(pheno_rats %>% filter(gene_name %in% body_fat_humans ) %>% .[["epifat"]],pch='+', col=3)

qqunif(pheno_rats$parafat,main="Parafat  Enrichment Plot in Rats")
qqpoints(pheno_rats %>% filter(gene_name %in% body_fat_humans ) %>% .[["parafat"]],pch='+', col=3)


qqunif(pheno_rats$retrofat,main="Retrofat  Enrichment Plot in Rats")
qqpoints(pheno_rats %>% filter(gene_name %in% body_fat_humans ) %>% .[["retrofat"]],pch='+', col=3)

```


```{r r flipped enrichment plot for height, obesity, BMI}
qqunif(pheno_rats$bodylength_w_tail,main="Bodylength with Tail  Enrichment Plot in Rats")
height_humans <- height_genes %>% filter(pvalue <= 0.05) %>% .[["gene_name"]] %>% na.omit() 
qqpoints(pheno_rats %>% filter(gene_name %in% height_humans ) %>% .[["bodylength_w_tail"]],pch='+', col=3)

qqunif(pheno_rats$bodylength_wo_tail,main="Bodylength without Tail  Enrichment Plot in Rats")
qqpoints(pheno_rats %>% filter(gene_name %in% height_humans ) %>% .[["bodylength_wo_tail"]],pch='+', col=3)

qqunif(pheno_rats$tail_length,main="Bodylength without Tail  Enrichment Plot in Rats")
qqpoints(pheno_rats %>% filter(gene_name %in% height_humans ) %>% .[["tail_length"]],pch='+', col=3)


qqunif(pheno_rats$bodyweight,main="Bodyweight Enrichment Plot in Rats")
obesity_humans <- obesity_genes %>% filter(pvalue <= 0.01) %>% .[["gene_name"]] %>% na.omit() 
qqpoints(pheno_rats %>% filter(gene_name %in% obesity_humans ) %>% .[["bodyweight"]],pch='+', col=3)


qqunif(pheno_rats$fasting_glucose,main="Fasting Glucose Enrichment Plot in Rats")
glucose_humans <- glucose_genes %>% filter(pvalue <= 0.01) %>% .[["gene_name"]] %>% na.omit() 
glucose_humans <- glucose_genes %>% filter(pvalue <= 0.01 && rcp >= 0.01) %>% .[["gene_name"]] %>% na.omit() 
qqpoints(pheno_rats %>% filter(gene_name %in% glucose_humans ) %>% .[["fasting_glucose"]],pch='+', col=3)

qqunif(pheno_rats$bmi_bodylength_w_tail,main="BMI Enrichment Plot in Rats")
BMI_humans <- BMI_genes %>% filter(pvalue <= 0.01) %>% .[["gene_name"]] %>% na.omit() 
qqpoints(pheno_rats %>% filter(gene_name %in% BMI_humans ) %>% .[["bmi_bodylength_w_tail"]],pch='+', col=3)

qqunif(pheno_rats$bmi_bodylength_wo_tail,main="BMI without tail Enrichment Plot in Rats")
qqpoints(pheno_rats %>% filter(gene_name %in% BMI_humans ) %>% .[["bmi_bodylength_wo_tail"]],pch='+', col=3)
```





