---
title: "4.PTRS_Creation"
author: "Natasha Santhanam"
date: "2/7/2022"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(data.table)
library(readxl)
library(qqman)
library(arrow)
library(RSQLite)
library(glmnet)
library(GenomicRanges)
library(liftOver)
"%&%" = function(a,b) paste(a,b,sep="")
dir <- "/Users/natashasanthanam/Github/rat-genomic-analysis/data/"
devtools::source_gist("ee5f67abddd0b761ee24410ea71c41aa")
```

```{r file dir, eval=FALSE}
data.dir <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/"
orth.rats <- read_tsv(data.dir %&% "expression/ortholog_genes_rats_humans.tsv", col_names = TRUE)
```

# Individual PTRS creation

Read in Weights and Expression and do some data wrangling
```{r read in weights and expression, eval=FALSE}
weights <- read_tsv(data.dir %&% "PTRS_weights/weight_files/elastic_net_alpha_0.1_British.export_model/weights.height.tsv.gz")
#weights <- read_tsv(data.dir %&% "PTRS_weights/weight_files/elastic_net_alpha_0.1_British.export_model/weights.bmi.tsv.gz")
weights$gene_id <- sapply(strsplit(weights$gene_id, "\\."), `[`, 1)

# Here we use predicted expression not predicted and make sure to use human ensembl id gene name
pred_expr <- fread(data.dir %&% "prediXcan/metabolic_traits/rat_metabolic_Ac__predict.txt") %>% select(-c(FID))

#filter only for genes that have a human ortholog
pred_expr <- pred_expr %>% select(c(IID, intersect(colnames(pred_expr), orth.rats$rnorvegicus_homolog_ensembl_gene) ))


#change name to human ensembl id in humans
colnames(pred_expr)[2:ncol(pred_expr)] <- orth.rats[match(colnames(pred_expr)[2:ncol(pred_expr)], orth.rats$rnorvegicus_homolog_ensembl_gene), 1] %>% .[["ensembl_gene_id"]]
  
rownames(weights) <- weights$gene_id
```

```{r generate predicted trait, eval=FALSE}
pred_trait <- fn_generate_trait(pred_expr, weights)
```

Compare all values across different models to actual BMI and Height
```{r read in pred trait}
pred_height <- readRDS(dir %&% "rat_pred_height_w_Human_PTRS.RDS") 
pred_BMI <- readRDS(dir %&% "rat_pred_bmi_w_Human_PTRS.RDS")

all_rats <- read_tsv(dir %&% "all_names.txt", col_names = TRUE)

pheno <- read_csv(dir %&% "processed_obesity_rat_Palmer_phenotypes.csv") %>% dplyr::select(c(rat_rfid, bmi_bodylength_w_tail, bmi_bodylength_wo_tail, bodylength_w_tail, bodylength_wo_tail, tail_length)) 
pheno <- pheno %>% filter(!(rat_rfid  %in% all_rats$ID))
```

Remove overlap of rats that were made in Prediction Models  - about 66 rats that overlap
```{r}
pred_height <- pred_height[-na.omit(match(all_rats$ID, rownames(pred_height))), ]
pred_BMI <- pred_BMI[-na.omit(match(all_rats$ID, rownames(pred_BMI))), ]
```

```{r check number of genes in each model}
weights_bmi <- fread("/Users/natashasanthanam/Downloads/weights.bmi.tsv.gz") 
weights_height <-  fread("/Users/natashasanthanam/Downloads/weights.height.tsv.gz")  

n_genes_bmi = as.matrix(apply(weights_bmi[,2:ncol(weights_bmi)], 2, function(x) sum(x != 0 )))
n_genes_height = as.matrix(apply(weights_height[,2:ncol(weights_height)], 2, function(x) sum(x != 0 )))
```


#Create Dataframes with the correlation coefficient between trait in rats and ones predicted using PTRS from Humans

BMI with predicted BMI
```{r check correlation for BMI }
bmi_with_tail <- pheno %>% dplyr::select(c(rat_rfid, bmi_bodylength_w_tail)) %>% na.omit()
tempo <- pred_BMI[na.omit(match(bmi_with_tail$rat_rfid, rownames(pred_BMI))), ]

bmi_w_tail_df <- data.frame(estimate = numeric(), pvalue = numeric(), model = character(), n_genes_in_model = numeric(), conf.int.min = numeric(), conf.int.max = numeric())
for(i in 1:ncol(tempo)){
  fit = lm(bmi_with_tail$bmi_bodylength_w_tail ~ tempo[,i])
  bmi_w_tail_df[i,1] <- summary(fit)$r.squared
  bmi_w_tail_df[i,2] <- with(summary(fit), pf(fstatistic[1],fstatistic[2],fstatistic[3],lower.tail=F))
  bmi_w_tail_df[i,3] <- paste("model", i, sep = "_")
  bmi_w_tail_df[i,4] <- n_genes_bmi[i,1]
  bmi_w_tail_df[i,5] <- confint(fit)[1] 
  bmi_w_tail_df[i,6] <- confint(fit)[2]
}


bmi_without_tail <- pheno %>% dplyr::select(c(rat_rfid, bmi_bodylength_wo_tail)) %>% na.omit()
tempo <- pred_BMI[na.omit(match(bmi_without_tail$rat_rfid, rownames(pred_BMI))), ]
bmi_wo_tail_df <- data.frame(estimate = numeric(), pvalue = numeric(),  model = character(), n_genes_in_model = numeric(), conf.int.min = numeric(), conf.int.max = numeric())
for(i in 1:ncol(tempo)){
  fit = lm(bmi_without_tail$bmi_bodylength_wo_tail ~ tempo[,i])
  bmi_wo_tail_df[i,1] <- summary(fit)$r.squared
  bmi_wo_tail_df[i,2] <- with(summary(fit), pf(fstatistic[1],fstatistic[2],fstatistic[3],lower.tail=F))
  bmi_wo_tail_df[i,3] <- paste("model", i, sep = "_")
  bmi_wo_tail_df[i,4] <- n_genes_bmi[i,1]
  bmi_wo_tail_df[i,5] <- confint(fit)[1] 
  bmi_wo_tail_df[i,6] <- confint(fit)[2] 
}

```

Bodylength with Predicted Height
```{r check correlation for height}
#Bodylength wit Tail vs predicted Height from Human PTRS weights
bodylength_w_tail <- pheno %>% dplyr::select(c(rat_rfid, bodylength_w_tail)) %>% na.omit()
tempo <- pred_height[na.omit(match(bodylength_w_tail$rat_rfid, rownames(pred_height))), ]

bodylength_w_tail_df <- data.frame(estimate = numeric(), pvalue = numeric(), model = character(), n_genes_in_model = numeric(), conf.int.min = numeric(), conf.int.max = numeric())
for(i in 1:ncol(tempo)){
  fit = lm(bodylength_w_tail$bodylength_w_tail ~ tempo[,i])
  bodylength_w_tail_df[i,1] <- summary(fit)$r.squared
  bodylength_w_tail_df[i,2] <- with(summary(fit), pf(fstatistic[1],fstatistic[2],fstatistic[3],lower.tail=F))
  bodylength_w_tail_df[i,3] <- paste("model", i, sep = "_")
  bodylength_w_tail_df[i,4]<- n_genes_height[i,1]
  bodylength_w_tail_df[i,5] <- confint(fit)[1] 
  bodylength_w_tail_df[i,6] <- confint(fit)[2] 
}


#now for bodylength without tail vs predicted Height 
bodylength_wo_tail <- pheno %>% dplyr::select(c(rat_rfid, bodylength_wo_tail)) %>% na.omit()
tempo <- pred_height[na.omit(match(bodylength_wo_tail$rat_rfid, rownames(pred_height))), ]

bodylength_wo_tail_df <- data.frame(estimate = numeric(), pvalue = numeric(),model = character(), n_genes_in_model = numeric(), conf.int.min = numeric(), conf.int.max = numeric())
for(i in 1:ncol(tempo)){
   fit = lm(bodylength_wo_tail$bodylength_wo_tail ~ tempo[,i])
  bodylength_wo_tail_df[i,1] <- summary(fit)$r.squared
  bodylength_wo_tail_df[i,2] <- with(summary(fit), pf(fstatistic[1],fstatistic[2],fstatistic[3],lower.tail=F))
  bodylength_wo_tail_df[i,3] <- paste("model", i, sep = "_")
  bodylength_wo_tail_df[i,4]<- n_genes_height[i,1]
  bodylength_wo_tail_df[i,5] <- confint(fit)[1] 
  bodylength_wo_tail_df[i,6] <- confint(fit)[2] 
}

#Tail-length vs predicted Height 
taillength <- pheno %>% dplyr::select(c(rat_rfid, tail_length)) %>% na.omit()
tempo <- pred_height[na.omit(match(taillength$rat_rfid, rownames(pred_height))), ]

tail_length_df <- data.frame(estimate = numeric(), pvalue = numeric(), model = character(), n_genes_in_model = numeric(), conf.int.min = numeric(), conf.int.max = numeric())
for(i in 1:ncol(tempo)){
  fit = lm(taillength$tail_length ~ tempo[,i])
  tail_length_df[i,1] <- summary(fit)$r.squared
  tail_length_df[i,2] <- with(summary(fit), pf(fstatistic[1],fstatistic[2],fstatistic[3],lower.tail=F))
  tail_length_df[i,3] <- paste("model", i, sep = "_")
  tail_length_df[i,4]<- n_genes_height[i,1]
  tail_length_df[i,5] <- confint(fit)[1] 
  tail_length_df[i,6] <- confint(fit)[2] 
}
```

# Plot Results

```{r plot pred height grapsh}
ggplot(tail_length_df, aes(n_genes_in_model, estimate)) + geom_point(color = "magenta", position="jitter") +geom_line(color = "magenta") + xlab("Number of genes in each model") + ylab("Partial R2") + ggtitle("Performance vs Number of Genes in Each Model for Tail length in Rats vs Predicted Height")

ggplot(bodylength_w_tail_df, aes(n_genes_in_model, estimate)) + geom_point() +geom_line() + xlab("Number of genes in each model") + ylab("Partial R2") + ggtitle("Performance vs Number of Genes in Each Model for Bodylength with Tail in Rats vs Predicted Height")
```
```{r plot pred bmi grapsh}
ggplot(bmi_w_tail_df, aes(n_genes_in_model, estimate)) + geom_point(color = "cyan", position="jitter") +geom_line(color = "cyan") + xlab("Number of genes in each model") + ylab("Partial R2") + ggtitle("Performance vs Number of Genes in Each Model for BMI with Tail in Rats vs Predicted BMI")
```
