---
title: "09.PRS_Rats"
author: "Natasha Santhanam"
date: "5/4/2022"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(data.table)
library(readxl)
library(qqman)
library(arrow)
library(RSQLite)
library(glmnet)
library(GenomicRanges)
library(liftOver)
library(ggpubr)
"%&%" = function(a,b) paste(a,b,sep="")
dir <- "/Users/natashasanthanam/Github/rat-genomic-analysis/data/"
geno.dir <- "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/"
```

## Run PRSice in Rats 

# Run GWAS in rats for bodylength

first extract out 500 random rats that we'll use as baseline
```{r subsampel 500 random rats}
fam <- fread(geno.dir %&% "Box_files/rat_genotypes_LD_pruned_0.95/plink_format/rat_metabolic_impute.fam")
all_names <- fread(geno.dir %&% "Box_files/all_names.txt")

fam <- fam %>% filter(!(V2 %in% all_names$ID))
base_ref <- data.frame(FID = sample(fam$V2, 500)) %>% mutate(ID = FID)
write_tsv(base_ref, geno.dir %&% "PTRS_weights/PRSice_Rats/PRSice_rats_reference_names.txt", col_names = FALSE)
```


```{bash generate plink files, eval=FALSE}
# remove any overlap of training rats
export DIR=/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/PTRS_weights/PRSice_Rats/
plink --bfile rat_metabolic_impute --remove $DIR/all_training_rats.txt  --make-bed --out rat_metabolic_abrv

#remove 500 rats that will later be used in baseline reference
plink --bfile rat_metabolic_abrv --remove $DIR/PRSice_rats_reference_names.txt --make-bed --out $DIR/rat_metabolic_PRSice

#create plink files for reference
plink --bfile rat_metabolic_abrv --keep $DIR/PRSice_rats_reference_names.txt --make-bed --out $DIR/rat_reference_PRSice
```

```{r create phenotype file, eval=FALSE}
fam <- read.table(DIR %&% "/rat_metabolic_PRSice.fam")
pheno <- read_tsv(geno.dir %&% "Box_files/processed_obesity_rat_Palmer_phenotypes.tsv") %>% select(c(rat_rfid, bodylength_w_tail))
pheno_all <- pheno[na.omit(match(fam$V1, pheno$rat_rfid)),]

pheno_all <- pheno_all %>% mutate(ID = rat_rfid, .before = "rat_rfid")
write_tsv(pheno_all, DIR %&% "bodylength_phenotype_PRSice.phe", col_names = FALSE)

pheno_ref <- pheno[na.omit(match(base_ref$FID, pheno$rat_rfid)),] 
pheno_ref <- pheno_ref %>% mutate(IID = pheno_ref$rat_rfid, .before = "rat_rfid")
write_tsv(pheno_ref, DIR %&% "bodylength_phenotype_ref_PRSice.phe", col_names = FALSE)
```

```{bash run GWAS, eval=FALSE}
plink --bfile rat_metabolic_PRSice --pheno bodylength_phenotype_PRSice.phe --assoc --out PRSIce_bodylength_assoc
```

# Run PRSice
```{r add effect allele to GWAS}
gwas <- fread(geno.dir %&% "PTRS_weights/PRSice_Rats/PRSIce_bodylength_assoc.qassoc")
bim <- fread(geno.dir %&% "PTRS_weights/PRSice_Rats/rat_metabolic_PRSice.bim")

#table(bim$V2 == gwas$SNP)
gwas <- gwas %>% mutate(A1 = bim$V5, .before = "BP")
write_tsv(gwas, "/gpfs/data/im-lab/nas40t2/natasha/rat_genomics/PTRS_weights/PRSice_Rats/PRSIce_bodylength_assoc_imputed.text",  col_names = TRUE)
```



```{bash}
 PRSice --base PRSIce_bodylength_assoc_imputed.text \
    --snp SNP \
    --chr CHR \
    --bp BP \
    --pvalue P \
    --A1 A1 \
    --binary-target F \
    --beta \
    --stat BETA \
    --target rat_reference_PRSice  \
    --pheno bodylength_phenotype_ref_PRSice.phe \
    --out rat_metabolic_PRSice
```


# Evaluate performance of PRSice

Plot performance as a function of pvalue threshold
```{r plot perf of PRSice as function of pval}
prsice_perf <- read_tsv(dir %&% "rat_metabolic_PRSice.prsice", col_names = TRUE)
ggplot(prsice_perf, aes(-log10(Threshold), R2)) + geom_point()
```
Check Correlation between observed bodylength and PRS from PRSice

```{r read in weights and expression for Framingham, eval=FALSE}
prs <- fread(dir %&% "rat_metabolic_PRSice.best")

all_rats <- read_tsv(dir %&% "all_names.txt", col_names = TRUE)
pheno <- read_csv(dir %&% "processed_obesity_rat_Palmer_phenotypes.csv") %>% dplyr::select(c(rat_rfid, bodylength_w_tail))
pheno <- pheno %>% filter(!(rat_rfid  %in% all_rats$ID))
```

```{r corrleation with PRSice, eval=FALSE}
pheno_height <- pheno[na.omit(match(prs$FID, pheno$rat_rfid)),] %>% na.omit()

tempo <- prs[na.omit(match(pheno$rat_rfid, prs$FID)),]
tempo <- tempo[match(pheno_height$rat_rfid, tempo$FID), ]

PRSice_height <- data.frame(estimate = numeric(), pvalue = numeric(), conf.int.min = numeric(), conf.int.max = numeric() )
 
  PRSice_height[1,1] <- cor.test(pheno_height$bodylength_w_tail, tempo$PRS)$estimate
  PRSice_height[1,2] <- cor.test(pheno_height$bodylength_w_tail, tempo$PRS)$p.value
  PRSice_height[1,3] <- cor.test(pheno_height$bodylength_w_tail, tempo$PRS)$conf.int[1]
  PRSice_height[1,4] <- cor.test(pheno_height$bodylength_w_tail, tempo$PRS)$conf.int[2]
```
